West runners dependencies in sync with CMake runners target

tejlmand opened the pull request:
This commit creates a phony `west_flash_target` target with identical
dependencies as CMake flash target (and the same for debug, debugserver,
etc. targets).

`ninja flash` correctly ensures dependencies are taken into
consideration before calling `west flash`.

Unfortunately, calling `west flash` directly would not re-run the flash
dependencies, cause `west flash` would only build the default CMake
target.

Now, `west flash` calls the phony `west_flash_depends` target, ensuring
all deps are up-to-date before flashing (unless --skip-rebuild is given)

hakonfam commented on scripts/west_commands/run_common.py:
```diff
@@ -363,17 +372,18 @@ def dump_context(command, args, unknown_args):
         cls = None
 
-    if runner_config is None:
+    if not runner_configs:
```
Personal opinion, feel free to ignore: Consider `if len(runner_configs) == 0` as it is more explicit and easier to read in my opinion, as you clearly see that its a list being checked.

hakonfam commented on scripts/west_commands/run_common.py:
```diff
@@ -340,14 +347,16 @@ def dump_traceback():
         board = cache['CACHED_BOARD']
         runners_yaml = runners_yaml_path(cache)
-        runner_config = load_runners_yaml(runners_yaml, args)
+        for yaml in runners_yaml:
+            runner_configs.append(load_runners_yaml(yaml, args))
```
```suggestion
        runner_configs = [load_runners_yaml(x, args) for x in runners_yaml]
```

tejlmand replied:
thanks

hakonfam commented on scripts/west_commands/run_common.py:
```diff
@@ -340,14 +347,16 @@ def dump_traceback():
+    runner_configs = []
     if build_dir is None:
         log.wrn('no --build-dir given or found; output will be limited')
         runner_config = None
```
```suggestion
        runner_configs = []
```

Either do this and remove the declaration above, or remove this declaration, because currently `runner_config` is left as an unused variable.

tejlmand replied:
removed.

hakonfam commented on scripts/west_commands/run_common.py:
```diff
@@ -363,17 +372,18 @@ def dump_context(command, args, unknown_args):
+            if cls:
+                dump_runner_context(command, cls, runner_config)
+            else:
+                dump_all_runner_context(command, runner_config, board, build_dir)
```
```suggestion
           dump_all_runner_context(command, runner_config, board,
                         build_dir)

```

Borked the indentation, but you get the idea, we need to break the line.

tejlmand replied:
fixed.

tejlmand commented on cmake/flash/CMakeLists.txt:
```diff
@@ -77,10 +77,15 @@ function(create_runners_yaml)
   file(GENERATE OUTPUT "${runners_yaml}" CONTENT
     $<TARGET_PROPERTY:runner_yml_props_target,yaml_contents>)
-  set(ZEPHYR_RUNNERS_YAML "${runners_yaml}" CACHE INTERNAL
+  get_property(EXTRA_ZEPHYR_RUNNERS_YAML
```
The reason for `EXTRA_ZEPHYR_RUNNERS_YAML` to be upper case, is because existing `ZEPHYR_RUNNERS_YAML` is upper case, and those two local variables are belonging together.
The reason for `ZEPHYR_RUNNERS_YML` property on `zephyr_property_target` to be uppercase, is because properties on targets are usually upper case:
https://cmake.org/cmake/help/latest/manual/cmake-properties.7.html#properties-on-targets

tejlmand commented on cmake/flash/CMakeLists.txt:
```diff
@@ -77,10 +77,15 @@ function(create_runners_yaml)
     endif()
   endforeach()
 
+  get_property(runners_yaml_extra GLOBAL PROPERTY ZEPHYR_MODULE_RUNNERS_YAML)
```
which code :see_no_evil:

hakonfam replied:
`runners_yaml_extra` I don't see it in this diff, is it used elsewhere in the code?

tejlmand replied:
no, it was leftover code, that I had forgot to remove.

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
+        runner_cls.add_parser(parser)
+        args, unknown = parser.parse_known_args(args=final_argv)
+        if unknown:
+            log.die(f'runner {runner_name} received unknown arguments: {unknown}')
```
never indent existing code........

tejlmand replied:
fixed

tejlmand commented on cmake/flash/CMakeLists.txt:
```diff
@@ -199,7 +199,6 @@ foreach(target flash debug debugserver attach)
       ${target}
       --skip-rebuild
       DEPENDS ${FLASH_DEPS}
-              $<TARGET_PROPERTY:zephyr_property_target,FLASH_DEPENDENCIES>
```
good point.
Done.
(commit message updated accordingly)

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
-                                cache)
-    runner_name = runner_cls.name()
-
+    # Load runners.yaml files.
```
yep it did.
Moved some code around, and apparently forgot the belonging comment

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
-        else:
-            log.err('verbose mode enabled, dumping stack:', fatal=True)
-            raise
+    for yaml in runners_yaml:
```
good point, although it has no impact in this case, then shadowing should be avoided.

tejlmand commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
+        runner_cls = use_runner_cls(command, board, user_args, runner_config,
+                                    cache)
+        runner_name = runner_cls.name()
+
```
fixed

mbolivar-nordic commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
-            log.err('verbose mode enabled, dumping stack:', fatal=True)
-            raise
+    for current_yaml in runners_yaml:
+        # A CMake list may contain multiple `;;` which can result in an empty
```
This is being called from dump_context() too, but the filtering is not being done there. I think the filtering should be done in runners_yaml_path() instead to make sure it always happens.

```
def runners_yaml_path(cache):
    ret = cache.get_list('ZEPHYR_RUNNERS_YAML')

    if ret is None:
        board = cache.get('CACHED_BOARD')
        log.die(f'either a pristine build is needed, or board {board} '
                "doesn't support west flash/debug "
                '(no ZEPHYR_RUNNERS_YAML in CMake cache)')

    # A CMake list may contain multiple `;;` which can result in an empty
    # entry in the result. Thus, simply ignore empty elements using filter().
    return [yml for yml in ret if yml]
```

This is also switching to use get_list(). Without that, this didn't work for me using a vanilla zephyr build, where the result is a string.

tejlmand replied:
Will move.

mbolivar-nordic replied:
Just making sure you noticed it's not just a move; there is another semantic change:

> This is also switching to use get_list(). Without that, this didn't work for me using a vanilla zephyr build, where the result is a string.

mbolivar-nordic reviewed (changes requested):
Noticed one issue during testing.

Thinking about this, I'm not really sure what the meaning of the `-r` option is anymore if different runners.yaml files support different runners. @tejlmand can you describe what you think ought to be done in that case, if e.g. I specify `-r openocd` and one `runners.yaml` has an `openocd` configuration but the other doesn't?

Another problem is how to handle tool-specific arguments... if we're turning this into a loop, we can't pass nrfjprog arguments sensibly if they only apply to one core.

mbolivar-nordic commented on scripts/west_commands/run_common.py:
```diff
@@ -142,90 +142,97 @@ def do_run_common(command, user_args, user_runner_args):
+
+        # Get a concrete ZephyrBinaryRunner subclass to use based on
+        # runners.yaml and command line arguments.
+        runner_cls = use_runner_cls(command, board, user_args, runner_config,
```
Testing this a bit more, even in the single-yaml-file case, I think it makes sense to pass use_runner_cls() the current_yaml value, so it can be logged:

```
def use_runner_cls(command, board, args, runner_config, cache, yaml_file):
...
    _banner(f'west {command.name}: using runner {runner} ({yaml_file})')
```

tejlmand replied:
good point and I agree.
Will add dedicated commit for that improvement.

tejlmand commented:
> Thinking about this, I'm not really sure what the meaning of the -r option is anymore if different runners.yaml files support different runners. @tejlmand can you describe what you think ought to be done in that case, if e.g. I specify -r openocd and one runners.yaml has an openocd configuration but the other doesn't?

Today there is only support for a single runner so I would argue that the enhancements in this PR for supporting multiple runners.yml files requires those to support the same runners.
So if `a/runners.yml` supports openocd and JLink, then `b/runners.yml` is also required to support openocd and JLink.

This being an enhancement to existing solution, then I think the requirement that all runners.yml must support the same runners is acceptable.
If people in future want even more possibilities, then it can be discussed how to cover those use-cases.

tejlmand commented:
> Another problem is how to handle tool-specific arguments... if we're turning this into a loop, we can't pass nrfjprog arguments sensibly if they only apply to one core.

This is a valid concern.
If user knows that only one core should have this argument, I would expect such user to be quite familiar with what he is doing.

So what about being able to specify a `runners.yml` on the command line to `west flash` ?
That way user has full control, like:
`west flash --runners-yaml a-core/runners.yml -- <tool-specific-flags for a-core>` 
`west flash --runners-yaml b-core/runners.yml `

tejlmand commented:
> > So what about being able to specify a runners.yml on the command line to west flash ?
> 
> It makes sense to do this regardless of whether we allow multiple runners.yaml files or just one, and in fact is something I wanted to do anyway already, but I think it doesn't solve this problem.
> 
> The advantage of having a `--runners-yaml` argument is that it would more easily allow you to use `west flash` independently of the build system: all you need is a runners.yaml file and the build artifacts it mentions inside, which could be specified using relative paths instead of absolute ones. You can zip that up and send it to someone, and they can flash the results without requiring a CMake cache.
> 

Seems we have a plan forward.

> Now that you're warming me up to the subject, I guess runners.yaml parsing should probably move into the runners package itself instead of run_common.py, now that we have more time to design this (I threw this format into v2.2 quickly for other reasons and haven't had time to revisit).
> 

You want to wait with this PR until that is in place, or could we continue this PR, and take that as followup work ?

> > That way user has full control, like:
> > `west flash --runners-yaml a-core/runners.yml -- <tool-specific-flags for a-core>`
> > `west flash --runners-yaml b-core/runners.yml `
> 
> Since I think this makes sense independently, I would argue that adding support for multiple such files would mean that `--runners-yaml` should be something that could be specified multiple times instead of just once, using `action='append'` when calling `ArgumentParser.add_argument`.
> 
> So this problem still remains.
> 

You mean the issue with passing additional flags to the runner ?
If we support multiple runners using a `--runners-yaml` and user specifies `--runners-yaml a/runners,.yml --runners.yml b/runners.yml -- <extra-arg>`, then `<extra-arg>` is added to ALL runners.
If user doesn't want that, then he must do:
```
west flash --runners-yaml a/runners,.yml  -- <extra-arg>
west flash --runners.yml b/runners.yml
```

basically:
```
west flash --runners-yaml a/runners,.yml  --runners.yml b/runners.yml -- <extra-arg>
# <extra-arg> on all runners
```
or
```
west flash --runners-yaml a/runners,.yml  -- <extra-arg>
west flash --runners.yml b/runners.yml
# <extra-arg> on only  a/runners.yml.
```

> As we've discussed, having extra dependencies on the `flash` target was done for a downstream-only purpose but it violates assumptions we have documented [here](https://docs.zephyrproject.org/latest/guides/west/without-west.html):
> 

Actually not, cause `west flash` (in shell) is not supported in that setup, but `ninja flash` will work if user follows the guide.
So setting dependencies in CMake in the flash target will actually work.
In the downstream use-case, then the dependency is to another `ninja flash` target, and if `ninja flash` works for main CMake flash target (if user has followed https://docs.zephyrproject.org/latest/guides/west/without-west.html), then it also works for the dependency (if that is another `ninja flash`).
So I see no violation. Can you explain why you see a violation ?

> > Running build system targets like ninja flash, ninja debug, etc. is just a call to the corresponding west command. For example, ninja flash calls west flash. If you don’t have west installed on your system, running those targets will fail.
> 

If it fails, even when users follow the guide, then it is not because of this PR, which simply introduces multiple runners.
But it must fail due to other changes.

> So now I am pretty convinced we should revert it at the very least, independently of whether we want to support multiple runners.yaml files in west flash.

Need some more explanation why you think the flash dependencies itself breaks it ?

mbolivar-nordic commented:
Torsten Tejlmand Rasmussen <notifications@github.com> writes:

>> > So what about being able to specify a runners.yml on the command line to west flash ?
>> 
>> It makes sense to do this regardless of whether we allow multiple runners.yaml files or just one, and in fact is something I wanted to do anyway already, but I think it doesn't solve this problem.
>> 
>> The advantage of having a `--runners-yaml` argument is that it would more easily allow you to use `west flash` independently of the build system: all you need is a runners.yaml file and the build artifacts it mentions inside, which could be specified using relative paths instead of absolute ones. You can zip that up and send it to someone, and they can flash the results without requiring a CMake cache.
>> 
>
> Seems we have a plan forward.

As discussed via chat, yep, looks like we'll go another way.

>
>> Now that you're warming me up to the subject, I guess runners.yaml parsing should probably move into the runners package itself instead of run_common.py, now that we have more time to design this (I threw this format into v2.2 quickly for other reasons and haven't had time to revisit).
>> 
>
> You want to wait with this PR until that is in place, or could we
> continue this PR, and take that as followup work ?

Not for this PR, for sure.

>
>> > That way user has full control, like:
>> > `west flash --runners-yaml a-core/runners.yml -- <tool-specific-flags for a-core>`
>> > `west flash --runners-yaml b-core/runners.yml `
>> 
>> Since I think this makes sense independently, I would argue that adding support for multiple such files would mean that `--runners-yaml` should be something that could be specified multiple times instead of just once, using `action='append'` when calling `ArgumentParser.add_argument`.
>> 
>> So this problem still remains.
>> 
>
> You mean the issue with passing additional flags to the runner ?
> If we support multiple runners using a `--runners-yaml` and user specifies `--runners-yaml a/runners,.yml --runners.yml b/runners.yml -- <extra-arg>`, then `<extra-arg>` is added to ALL runners.
> If user doesn't want that, then he must do:
> ```
> west flash --runners-yaml a/runners,.yml  -- <extra-arg>
> west flash --runners.yml b/runners.yml
> ```
>
> basically:
> ```
> west flash --runners-yaml a/runners,.yml  --runners.yml b/runners.yml -- <extra-arg>
> # <extra-arg> on all runners
> ```
> or
> ```
> west flash --runners-yaml a/runners,.yml  -- <extra-arg>
> west flash --runners.yml b/runners.yml
> # <extra-arg> on only  a/runners.yml.
> ```

We've discussed another approach where we just preserve a single
runners.yaml.

>
>> As we've discussed, having extra dependencies on the `flash` target was done for a downstream-only purpose but it violates assumptions we have documented [here](https://docs.zephyrproject.org/latest/guides/west/without-west.html):
>> 
>
> Actually not, cause `west flash` (in shell) is not supported in that setup, but `ninja flash` will work if user follows the guide.
> So setting dependencies in CMake in the flash target will actually work.
> In the downstream use-case, then the dependency is to another `ninja flash` target, and if `ninja flash` works for main CMake flash target (if user has followed https://docs.zephyrproject.org/latest/guides/west/without-west.html), then it also works for the dependency (if that is another `ninja flash`).
> So I see no violation. Can you explain why you see a violation ?
>
>> > Running build system targets like ninja flash, ninja debug, etc. is just a call to the corresponding west command. For example, ninja flash calls west flash. If you don’t have west installed on your system, running those targets will fail.
>> 
>
> If it fails, even when users follow the guide, then it is not because of this PR, which simply introduces multiple runners.
> But it must fail due to other changes.
>
>> So now I am pretty convinced we should revert it at the very least, independently of whether we want to support multiple runners.yaml files in west flash.
>
> Need some more explanation why you think the flash dependencies itself
> breaks it ?

To summarize our offline discussion, we'll go with:

1. Add a new flash_deps target where we add any dependencies the build system recommends west to get up to date before flashing
2. "west flash" will run the flash_deps target instead of the default (all) target
3. "west flash --skip-rebuild" will avoid running this target, basically same as it does today, but just with a different target
4. the existing "flash" cmake target will depend on "flash_deps" and then call "west flash --skip-rebuild" directly
5. the "flash_deps" target by default will at least refresh ${logical_target_for_zephyr_elf} so backwards compatibility is maintained, but can also be extended

IOW, same as today but just a slight change of what "west flash"
rebuilds by default to restore the property that "ninja flash" is
basically just a "refresh dependencies and run west flash". Only the
name of the target used to refresh dependencies changes, so users can
start tacking on additional flash dependencies if desired. However, this
is not assumed and can be turned off with "west flash --skip-rebuild",
just like today.

>
> -- 
> You are receiving this because your review was requested.
> Reply to this email directly or view it on GitHub:
> https://github.com/zephyrproject-rtos/zephyr/pull/26365#issuecomment-649072046

tejlmand committed:
cmake: west: west flash dependencies in sync with CMake flash target

This commit creates a list of a phony targets for each runner, that is:
`west_flash_depends`, `west_debug_depends`, and so on.
Those targets has identical dependencies as CMake runner target.
flash, debug, debugserver, attach targets.

As example `ninja flash` correctly ensures dependencies are taken into
consideration before calling `west flash`.

Unfortunately, calling `west flash` directly would not re-run the flash
dependencies, cause `west flash` would only build the default CMake
target.

Now, `west flash` calls the phony `west_flash_depends` target, ensuring
all deps are up-to-date before flashing (unless --skip-rebuild is given)

The same is true for the other mentioned runners.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic commented on scripts/west_commands/run_common.py:
```diff
@@ -260,7 +260,7 @@ def load_cmake_cache(build_dir, args):
     _banner(f'west {command.name}: rebuilding')
     try:
-        zcmake.run_build(build_dir)
+        zcmake.run_build(build_dir, target="west_flash_depends")
```
This is shared code called by all the runner related commands, which includes things like debug, debugserver, and attach. It looks to me like this is assuming that it's only called for flash, even though the changes to cmake/flash/CMakeLists.txt do produce a separate dependency target for each command.

Shouldn't this be something like `target=f"west_{command.name}_depends"` instead, then?

mbolivar-nordic commented on scripts/west_commands/zcmake.py:
```diff
@@ -71,17 +71,20 @@ def run_cmake(args, cwd=None, capture_output=False, dry_run=False):
 
 
-def run_build(build_directory, **kwargs):
+def run_build(build_directory, target=None, **kwargs):
```
This is a little tricky in my opinion. Maybe we can just make run_build use extra_args instead the way build.py does? My concern is things like how extra_args in build.py adds additional arguments to the underlying build tool with `--`. If a caller did that *and* set the `target` kwarg, it would produce unexpected and incorrect results.

Not a blocker, just want to explore this alternative instead.

tejlmand replied:
I did consider to use the `extra_args` but I felt it was cleaner to have a dedicated parameter for this, and have the `extra_args` reserved for the command line.

> If a caller did that and set the target kwarg

Doing so could already today give unintended side effects, like if you do:
```
west flash -- --target menuconfig
```
that would not build the hex file.

In order to ensure (better) consistency between `west flash` and `ninja flash` (as well as other runners), I think it makes sense to reserve `--target <target>` when invoking `west flash`, as this PR does.

If user wants to build a dedicated target first and NOT have the flash deps target executed, then user can still do:
```
west build --target <mytarget>
west flash --skip-rebuild
```

mbolivar-nordic replied:
> Doing so could already today give unintended side effects, like if you do:
> 
> ```
> west flash -- --target menuconfig
> ```
> 
> that would not build the hex file.

I don't understand; `west flash` doesn't take extra command line arguments like this, does it?

tejlmand replied:
seems you're right, I though the `--extra-args` was for the command line of `west flash` being passed as kwarg when using `--`.

I see that I was wrong on that interpretation, and thus I completely agree, the kwarg should be used instead.

tejlmand replied:
fixed.

mbolivar-nordic commented on cmake/flash/CMakeLists.txt:
```diff
@@ -210,6 +210,16 @@ foreach(target flash debug debugserver attach)
+    # This allows `west` to run just the dependencies before flashing.
+    add_custom_target(west_${target}_depends
+      COMMAND ${CMAKE_COMMAND} -E echo ""
+      DEPENDS ${FLASH_DEPS}
```
Different runner commands are not going to want FLASH_DEPS, I think. Especially `attach`, which by contract doesn't do any reflashing. Or am I missing something here?

tejlmand replied:
No you're not, but today all runners are setup in a loop:
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/flash/CMakeLists.txt#L167
https://github.com/zephyrproject-rtos/zephyr/blob/master/cmake/flash/CMakeLists.txt#L201-L202
```
foreach(target flash debug debugserver attach)
...
     set(cmd
      ...
      ${target}
      --skip-rebuild
      DEPENDS ${FLASH_DEPS}
              $<TARGET_PROPERTY:zephyr_property_target,FLASH_DEPENDENCIES>
```
which means all runners has exactly the same dependencies, which for some of the runners makes sense.

The main goal of the updated PR is to make things consistent between ninja and west runners.

You can argue that `attach`, `debug`, `debugserver`, and `flash` should have individual / independent dependencies (`debug` and `flash` could possibly have same dependencies).
But I found that to be outside the scope of having `west <runner>` behave identical to `ninja <runner>`.
I will be happy to look into cleaning the dependencies of the runners also.

tejlmand replied:
>  I will be happy to look into cleaning the dependencies of the runners also.

Probably in separate PR, to keep things separated.

mbolivar-nordic replied:
> which means all runners has exactly the same dependencies, which for some of the runners makes sense.

TBH I'm worried about our downstream users. If we add dependencies which do some flashing of sub-images here in this way, won't `west attach` break for nRF53 in NCS?

mbolivar-nordic replied:
By "break", I mean "do some reflashing of sub-images, which its contract says it will not do".

tejlmand replied:
that is actually a good point, but it is not `west` that breaks this contract / this PR, but the dependency in CMake.

So `ninja attach` could have side effects already today.
Cause as shown in the link, the depends is already on `master`.

Will add an extra commit to this PR, to clean the dependencies.

mbolivar-nordic replied:
Sounds good. The main thing is that we have separate dependency targets for each command, which makes sense, so we can keep the dependencies for flash an debug (which may flash the target) separate from the ones for attach (which may not).

tejlmand replied:
fixed.
Also changed the old `FLASH_DEPS` into the more correctly named `RUNNERS_DEPS` during this cleanup.

tejlmand committed:
cmake: runners dependencies cleanup

This commit renames `FLASH_DEPS` to `RUNNERS_DEPS`.
The current name `FLASH_DEPS` is misleading in the sense that this
depency is added to all runner targets, flash, debug, debugserver,
attach, and not only the flash runners.

Therefore this is now named `RUNNERS_DEPS` instead.

Similar, zephyr_property_target now contains the property
FLASH_DEPENDENCIES, DEBUG_DEPENDENCIES, and so on, so that additional
dependencies can be added for each runner.

Signed-off-by: Torsten Rasmussen <Torsten.Rasmussen@nordicsemi.no>

mbolivar-nordic reviewed (approved):
I think this seems sane, thanks for all the work on this. I like that the final solution is small and extensible.

tejlmand commented:
> 
> 
> I think this seems sane, thanks for all the work on this. I like that the final solution is small and extensible.

Thanks for review and good discussions.
A lot of suggestions and ideas along the way, but I too like this final solution for the exact same reasons.

hakonfam reviewed (approved):
Nice
